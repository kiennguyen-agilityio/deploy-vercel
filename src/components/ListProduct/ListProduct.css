.product-list {
  display: flex;
  flex-wrap: wrap;
  margin-left: 70px;
}

.no-results {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  padding: 20px;
  text-align: center;
}

.no-results h2 {
  font-size: 1.5rem;
  margin-bottom: 10px;
}

.no-results button {
  background-color: var(--color-primary);
  color: white;
  border: none;
  padding: 10px 20px;
  border-radius: 4px;
  cursor: pointer;
  transition: background-color 0.3s ease;
}

.no-results button:hover {
  background-color: var(--color-primary-dark);
}
