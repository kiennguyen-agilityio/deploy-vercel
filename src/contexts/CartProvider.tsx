import { createContext, useMemo, useCallback, ReactNode, useReducer, useContext } from 'react';
import { mutate } from 'swr';

// constants
import {
  ADD_PRODUCT_FAILURE,
  DELETE_PRODUCT_FAILURE,
  UPDATE_PRODUCT_FAILURE,
  ADD_PRODUCT_REQUEST,
  ADD_PRODUCT_SUCCESS,
  DELETE_PRODUCT_REQUEST,
  DELETE_PRODUCT_SUCCESS,
  UPDATE_PRODUCT_REQUEST,
  UPDATE_PRODUCT_SUCCESS,
} from '@/constants/actionType';

// services
import { addToCart, deleteCartItem, updateItem } from '@/services/api-action';

// hooks
import { useCartProducts } from '@/components/hooks/fetch';

// constants
import { CART_API } from '@/constants/api';

// reducers
import cartReducer, { ProductCartState } from '@/reducers/cartReducer';

// interfaces
import Product from '@/interface/Product';

export const initialState: ProductCartState = {
  error: '',
  isLoading: false,
  isAdding: false,
  products: [],
};

export interface CartContextProps {
  error: string;
  isAdding: boolean;
  isLoading: boolean;
  products?: Product[];
  onAddToCart: (product: Product) => void;
  onUpdateCart: (productId: string, updatedData: Partial<Product>, url: string) => Promise<void>;
  onDeleteProductFromCart: (productId: string) => void;
}

export const CartContext = createContext<CartContextProps>({
  error: '',
  isAdding: false,
  isLoading: false,
  products: [],
  onAddToCart: () => undefined,
  onUpdateCart: async () => {},
  onDeleteProductFromCart: async () => {},
});

interface CartProviderProps {
  children: ReactNode;
}

export const useProductCartContext = () => useContext(CartContext);

const CartProvider: React.FC<CartProviderProps> = ({ children }) => {
  const [state, dispatch] = useReducer(cartReducer, initialState);
  const { isLoading, error, isAdding } = state;
  const { data: productsFromSwr } = useCartProducts();

  const handleAddToCart = useCallback(async (product: Product): Promise<void> => {
    try {
      dispatch({ type: ADD_PRODUCT_REQUEST });

      const data = await addToCart(product);

      mutate(CART_API);

      dispatch({ type: ADD_PRODUCT_SUCCESS, payload: data });
    } catch {
      dispatch({ type: ADD_PRODUCT_FAILURE });
    }
  }, []);

  const handleDeleteProductFromCart = useCallback(async (productId: string): Promise<void> => {
    try {
      dispatch({ type: DELETE_PRODUCT_REQUEST });

      await deleteCartItem(productId);

      mutate(CART_API);

      dispatch({ type: DELETE_PRODUCT_SUCCESS, payload: productId });
    } catch (error) {
      dispatch({ type: DELETE_PRODUCT_FAILURE });
    }
  }, []);

  const handleUpdateItem = useCallback(
    async (productId: string, updatedData: Partial<Product>, url: string) => {
      try {
        dispatch({ type: UPDATE_PRODUCT_REQUEST });
        await updateItem(productId, updatedData, url);

        mutate(CART_API);

        const updatedProductData: Partial<Product> = {
          ...updatedData,
          id: productId,
        };

        dispatch({ type: UPDATE_PRODUCT_SUCCESS, payload: updatedProductData });
      } catch {
        dispatch({ type: UPDATE_PRODUCT_FAILURE });
      }
    },
    []
  );

  const cartContextValue = useMemo(
    () => ({
      isLoading,
      error,
      products: productsFromSwr || [],
      isAdding,
      onAddToCart: handleAddToCart,
      onUpdateCart: handleUpdateItem,
      onDeleteProductFromCart: handleDeleteProductFromCart,
    }),
    [state, productsFromSwr]
  );

  return <CartContext.Provider value={cartContextValue}>{children}</CartContext.Provider>;
};

export default CartProvider;
