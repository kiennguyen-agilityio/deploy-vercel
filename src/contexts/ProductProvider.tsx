import { createContext, ReactNode, useState, useMemo } from 'react';

// hooks
import { useProducts } from '@/components/hooks/fetch';

// interfaces
import Product from '@/interface/Product';

interface ProductProviderProps {
  children: ReactNode;
}

export interface ListProductContextProps {
  products?: Product[];
  selectedProduct?: Product | null;
  onGetProductById: (productId: string) => void;
  error: string;
  isLoading: boolean;
}

export const ListProductContext = createContext<ListProductContextProps>({
  onGetProductById: async () => {},
  error: '',
  isLoading: false,
});

export const ListProductProvider: React.FC<ProductProviderProps> = ({ children }) => {
  const { data: productsFromSwr, error: productsError, isLoading } = useProducts();
  const [selectedProduct, setSelectedProduct] = useState<Product | null>(null);

  const getProductById = (productId: string) => {
    const selectedProduct = productsFromSwr?.find((product) => product.id === productId);
    setSelectedProduct(selectedProduct || null);
  };

  const productContextValue = useMemo(
    () => ({
      products: productsFromSwr,
      selectedProduct,
      onGetProductById: getProductById,
      error: productsError ? productsError.message : '',
      isLoading,
    }),
    [productsError, productsFromSwr, selectedProduct]
  );

  return (
    <ListProductContext.Provider value={productContextValue}>
      {children}
    </ListProductContext.Provider>
  );
};
