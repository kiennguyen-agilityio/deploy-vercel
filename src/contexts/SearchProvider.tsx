import { createContext, useState, useMemo, ReactNode } from 'react';

export interface SearchContextProps {
  searchValue: string;
  onSearchValue: (value: string) => void;
  onResetSearchValue: () => void;
}
interface SearchProviderProps {
  children: ReactNode;
}

export const SearchContext = createContext<SearchContextProps>({} as SearchContextProps);

export const SearchProvider: React.FC<SearchProviderProps> = ({ children }) => {
  const [searchValue, setSearchValue] = useState<string>('');

  const resetSearchValue = (): void => {
    setSearchValue('');
  };

  const setValue = (value: string): void => {
    setSearchValue(value);
  };

  const contextValue = useMemo(
    () => ({
      searchValue,
      setSearchValue,
      onResetSearchValue: resetSearchValue,
      onSearchValue: setValue,
    }),
    [searchValue, setSearchValue]
  );

  return <SearchContext.Provider value={contextValue}>{children}</SearchContext.Provider>;
};
