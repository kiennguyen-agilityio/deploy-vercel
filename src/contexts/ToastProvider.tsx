import { createContext, useContext, useMemo, useState, ReactNode } from 'react';

interface ToastContextValue {
  toast: {
    status: string;
    message: string;
    openPopup: boolean;
  };
  setToast: React.Dispatch<
    React.SetStateAction<{
      status: string;
      message: string;
      openPopup: boolean;
    }>
  >;
}

const ToastContext = createContext<ToastContextValue | undefined>(undefined);

const useToast = (): ToastContextValue => {
  const context = useContext(ToastContext);
  if (!context) {
    throw new Error('useToast must be used within a ToastProvider');
  }

  return context;
};

interface ToastProviderProps {
  children: ReactNode;
}

const ToastProvider: React.FC<ToastProviderProps> = ({ children }) => {
  const [toast, setToast] = useState({
    status: '',
    message: '',
    openPopup: false,
  });

  const valueContext = useMemo(
    () => ({
      toast,
      setToast,
    }),
    [toast]
  );

  return <ToastContext.Provider value={valueContext}>{children}</ToastContext.Provider>;
};

export { useToast, ToastProvider };
