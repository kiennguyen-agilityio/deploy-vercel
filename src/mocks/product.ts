export const productList = [
  {
    id: 1,
    name: 'Fullset Black Chair & Sofa',
    sale: 360,
    price: 5000,
    discountPercent: 10,
    image: '../src/assets/images/full-black-chair-sofa.png',
    descriptions:
      'black sofa fullset, made of leather selection from the Himalayan mountains. Knitted by the professional hands of housewives. Making a masterpiece of the best sofa in the world. Very soft and comfortable',
    quantity: 2,
    color: 'Red',
    subcategory: 'Office chair',
  },
  {
    id: 2,
    name: ' Desk Dark Chair',
    sale: 400,
    price: 4000,
    discountPercent: 20,
    image: '../src/assets/images/soft-dark-chair.png',
    descriptions:
      'black sofa fullset, made of leather selection from the Himalayan mountains. Knitted by the professional hands of housewives. Making a masterpiece of the best sofa in the world. Very soft and comfortable',
    quantity: 3,
    color: 'Green',
    subcategory: 'Family chair',
  },
  {
    id: 3,
    name: 'Orange Color Chairs',
    sale: 400,
    price: 3000,
    discountPercent: 0,
    image: '../src/assets/images/orange-cool-chair.png',
    descriptions:
      'black sofa fullset, made of leather selection from the Himalayan mountains. Knitted by the professional hands of housewives. Making a masterpiece of the best sofa in the world. Very soft and comfortable',
    quantity: 3,
    color: 'Red',
    subcategory: 'Family chair',
  },
  {
    id: 4,
    name: 'Soft Dark Chair',
    sale: 400,
    price: 2000,
    discountPercent: 0,
    image: '../src/assets/images/full-orange-chair.png',
    descriptions:
      'black sofa fullset, made of leather selection from the Himalayan mountains. Knitted by the professional hands of housewives. Making a masterpiece of the best sofa in the world. Very soft and comfortable',
    quantity: 3,
    color: 'Yellow',
    subcategory: 'Reading chair',
  },
  {
    id: 5,
    name: 'Soft Dark Chair',
    sale: 400,
    price: 1500,
    discountPercent: 0,
    image: '../src/assets/images/blue-cabinet.png',
    descriptions:
      'black sofa fullset, made of leather selection from the Himalayan mountains. Knitted by the professional hands of housewives. Making a masterpiece of the best sofa in the world. Very soft and comfortable',
    quantity: 4,
    color: 'Black',
    subcategory: 'Reading chair',
  },
  {
    id: 6,
    name: 'Soft Dark Chair',
    sale: 400,
    price: 1400,
    discountPercent: 0,
    image: '../src/assets/images/black-chair.png',
    descriptions:
      'black sofa fullset, made of leather selection from the Himalayan mountains. Knitted by the professional hands of housewives. Making a masterpiece of the best sofa in the world. Very soft and comfortable',
    quantity: 2,
    subcategory: 'Reading chair',
  },
];
