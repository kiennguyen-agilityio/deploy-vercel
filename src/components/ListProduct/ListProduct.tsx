import { useContext, useState } from 'react';

// constants
import ERROR_MESSAGES from '@/constants/errorMsg';

// contexts
import { SearchContext } from '@/contexts/SearchProvider';
import { ListProductContext } from '@/contexts/ProductProvider';

// enums
import { SORT } from '@/enums/index';

// interface
import Product from '@/interface/Product';

// components
import ProductCard from '@/components/ProductCard/ProductCard';
import Pagination from '@/components/Pagination';
import LoadingSpinner from '@/components/Common/LoadingSpinner/LoadingSpinner';

// style
import './ListProduct.css';

export interface Props {
  selectedColors: string[];
  selectedPrice: number | null;
  sortType: 'asc' | 'desc' | '';
  selectedSubcategory?: string;
}

const ListProduct: React.FC<Props> = ({ selectedColors, sortType, selectedSubcategory }) => {
  const [currentPage, setCurrentPage] = useState<number>(1);
  const [productsPerPage] = useState<number>(15);
  const { searchValue } = useContext(SearchContext);
  const { onSearchValue } = useContext(SearchContext);

  const { products, isLoading } = useContext(ListProductContext);

  const filteredProducts =
    products?.filter((product) => {
      // Check if the product's color is included in the selected colors
      const isIncludedSelectedColors = selectedColors?.includes(product.color);

      // Check if the product's subcategory matches the selected subcategory
      const isIncludedSelectedSubcategory = selectedSubcategory === product.subcategory;

      // Convert the product name and search value to lowercase for case-insensitive comparison
      const productName = product.name.toLowerCase();
      const searchTerm = searchValue.trim().toLowerCase();

      // Return true if the product name contains the search term and meets the selected colors and subcategory criteria
      return (
        productName.includes(searchTerm) &&
        (!selectedColors.length || isIncludedSelectedColors) &&
        (!selectedSubcategory?.length || isIncludedSelectedSubcategory)
      );
    }) || [];

  /**
   * Sorts the products based on the given sort type.
   * @param {Array<Product>} filteredProducts - The array of filtered products.
   * @param {string} sortType - The sort type (ASC or DESC).
   * @returns {Array<Product>} - The sorted products array.
   */
  const sortedProducts = filteredProducts?.sort((a, b) => {
    // Sort the products based on the sort type
    if (sortType === SORT.ASC) {
      return a.price - b.price;
    }

    if (sortType === SORT.DESC) {
      return b.price - a.price;
    }

    return 0;
  });

  // Calculate the index of the last product on the current page
  const indexOfLastProduct = currentPage * productsPerPage;

  // Calculate the index of the first product on the current page
  const indexOfFirstProduct = indexOfLastProduct - productsPerPage;

  // Get the subset of products for the current page
  const currentProducts = sortedProducts.slice(indexOfFirstProduct, indexOfLastProduct);

  // Calculate the total number of pages based on the number of products and products per page
  const totalPages = Math.ceil(sortedProducts.length / productsPerPage);

  const handlePageChange = (pageNumber: number): void => {
    setCurrentPage(pageNumber);
  };

  return (
    <div>
      {isLoading && <LoadingSpinner />}
      <div className="product-list">
        {currentProducts.length > 0 ? (
          currentProducts.map((product: Product) => (
            <ProductCard key={product.id} product={product} />
          ))
        ) : (
          <div className="no-results">
            {searchValue ? ERROR_MESSAGES.NOT_FOUND : ''}
            {searchValue && (
              <button type="button" onClick={() => onSearchValue('')}>
                Try again
              </button>
            )}
          </div>
        )}
      </div>
      <Pagination
        currentPage={currentPage}
        totalPages={totalPages}
        onPageChange={handlePageChange}
      />
    </div>
  );
};

export default ListProduct;
