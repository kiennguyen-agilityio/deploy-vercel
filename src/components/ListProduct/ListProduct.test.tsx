import { fireEvent, render, waitFor } from '@testing-library/react';
import axios from 'axios';

// contexts
import { LoadingProvider } from '@/contexts/LoadingProvider';

// mocks
import { productList } from 'mocks/product';

// components
import ListProduct, { Props } from './ListProduct';

const mockLoadingValue = {
  isLoading: false,
  setLoading: jest.fn(),
};

jest.mock('@/contexts/LoadingProvider', () => ({
  LoadingProvider: ({ children }: { children: React.ReactNode }) => children,
  useLoading: () => mockLoadingValue,
}));

const mockProps: Props = {
  selectedColors: [''],
  selectedPrice: 0,
  sortType: 'asc',
  selectedSubcategory: '',
};

const setup = (overrides = {}) => {
  mockLoadingValue.setLoading.mockReset();
  jest.spyOn(axios, 'get').mockResolvedValue({ data: { products: productList } });

  const props = { ...mockProps, ...overrides };
  return render(
    <LoadingProvider>
      <ListProduct {...props} />
    </LoadingProvider>
  );
};

describe('ProductList component', () => {
  beforeEach(() => {
    mockLoadingValue.setLoading.mockReset();
    jest.restoreAllMocks();
  });

  const mockProps: Props = {
    selectedColors: ['Red'],
    selectedPrice: 2000,
    sortType: 'asc',
    selectedSubcategory: 'Office chair',
  };

  it('List renders correctly', () => {
    const { container } = setup(mockProps);

    console.log(container.innerHTML);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render list of products when products are not empty', async () => {
    const { container, queryByText } = setup(mockProps);
    await queryByText('Fullset Black Chair & Sofa');
    expect(container.firstChild).toMatchSnapshot();
    expect(mockLoadingValue.setLoading).toHaveBeenCalledTimes(2);
  });

  it('should filter products correctly based on selected filters', () => {
    const { container } = setup(mockProps);
    expect(container.firstChild).toMatchSnapshot();
    expect(mockLoadingValue.setLoading).toHaveBeenCalledTimes(1);
  });

  it('should sort products correctly based on sort type', () => {
    const { container, queryAllByAltText } = setup(mockProps);
    const elements = queryAllByAltText('Maximum Price');

    elements.length > 0 && elements[0].click();

    expect(container.firstChild).toMatchSnapshot();
    expect(mockLoadingValue.setLoading).toHaveBeenCalledTimes(1);
  });

  it('should handle page change correctly', () => {
    const { container, getByText } = setup();

    waitFor(() => {
      const element = getByText('1');

      expect(element).toBeInTheDocument();
      fireEvent.click(element);
    });

    expect(container.firstChild).toMatchSnapshot();
    expect(mockLoadingValue.setLoading).toHaveBeenCalledTimes(1);
  });

  it('returns false when selectedColors has items and color is not included', () => {
    const { container } = setup({
      selectedColors: ['Red', 'Blue'],
      selectedPrice: 0,
      sortType: 'asc',
      selectedSubcategory: '',
    });

    const productCards = container.querySelectorAll('.product-card');
    expect(productCards.length).toBe(0);
  });

  it('returns products sorted in ascending order by price when sortType is "asc"', () => {
    const products = [
      { id: 1, price: 50 },
      { id: 2, price: 30 },
      { id: 3, price: 70 },
    ];

    const sortedProducts = products.sort((a, b) => a.price - b.price);

    expect(sortedProducts).toEqual([
      { id: 2, price: 30 },
      { id: 1, price: 50 },
      { id: 3, price: 70 },
    ]);
  });

  fit('returns products sorted in descending order by price when sortType is "desc"', () => {
    const products = [
      { id: 1, price: 50 },
      { id: 2, price: 30 },
      { id: 3, price: 70 },
    ];

    const sortedProducts = products.sort((a, b) => b.price - a.price);

    expect(sortedProducts).toEqual([
      { id: 3, price: 70 },
      { id: 1, price: 50 },
      { id: 2, price: 30 },
    ]);
  });
});
