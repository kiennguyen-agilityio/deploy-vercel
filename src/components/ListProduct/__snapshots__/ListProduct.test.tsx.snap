// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList component List renders correctly 1`] = `
<div>
  <div
    class="product-list"
  >
    <div
      class="no-results"
    />
  </div>
  <div
    class="pagination-container"
  >
    <nav
      aria-label="Pagination"
    >
      <ul
        class="pagination"
      />
    </nav>
  </div>
</div>
`;

exports[`ProductList component should filter products correctly based on selected filters 1`] = `
<div>
  <div
    class="product-list"
  >
    <div
      class="no-results"
    />
  </div>
  <div
    class="pagination-container"
  >
    <nav
      aria-label="Pagination"
    >
      <ul
        class="pagination"
      />
    </nav>
  </div>
</div>
`;

exports[`ProductList component should handle page change correctly 1`] = `
<div>
  <div
    class="product-list"
  >
    <div
      class="no-results"
    />
  </div>
  <div
    class="pagination-container"
  >
    <nav
      aria-label="Pagination"
    >
      <ul
        class="pagination"
      />
    </nav>
  </div>
</div>
`;

exports[`ProductList component should render list of products when products are not empty 1`] = `
<div>
  <div
    class="product-list"
  >
    <div
      class="no-results"
    />
  </div>
  <div
    class="pagination-container"
  >
    <nav
      aria-label="Pagination"
    >
      <ul
        class="pagination"
      />
    </nav>
  </div>
</div>
`;

exports[`ProductList component should sort products correctly based on sort type 1`] = `
<div>
  <div
    class="product-list"
  >
    <div
      class="no-results"
    />
  </div>
  <div
    class="pagination-container"
  >
    <nav
      aria-label="Pagination"
    >
      <ul
        class="pagination"
      />
    </nav>
  </div>
</div>
`;
